vtkext/private/module/vtkF3DRenderer.cxx

```diff
@@ -290,59 +293,65 @@ void vtkF3DRenderer::InitializeUpVector(const std::vector<double>& upVec)
 {
   assert(upVec.size() == 3);
 
-  // TODO: Remove this when https://github.com/f3d-app/f3d/pull/1536/ is merged
-  int upIndex = -1;
-  bool error = true;
-  for (int i = 0; i < 3; i++)
+  const auto isNullVector = [](const std::array<double, 3>& v)
   {
-    if (std::abs(upVec[i]) == 1)
-    {
-      if (upIndex == -1)
-      {
-        upIndex = i;
-        error = false;
-      }
-      else
-      {
-        error = true;
-        break;
-      }
-    }
+    constexpr double e = 1e-8;
+    return ::abs(v[0]) < e && ::abs(v[1]) < e && ::abs(v[2]) < e;
+  };
+
+  std::array<double, 3> up = { upVec[0], upVec[1], upVec[2] };
+  std::array<double, 3> right = { 1, 0, 0 };
+
+  /* if `up` is `(0,0,0)` make it `(0,1,0)` */
+  if (isNullVector(up))
+  {
+    up[1] = 1.0;
+    F3DLog::Print(F3DLog::Severity::Warning, "null up vector, using (0,0,1) instead");
   }
+  vtkMath::Normalize(up.data());
 
-  if (!error)
+  /* make sure `right` is not colinear with `up` */
+  for (size_t i = 0; ::abs(vtkMath::Dot(right, up)) > 0.999 && i < 3; ++i)
```

Comment:
It feels like the for loop is not required. What about?
```cpp
if (::abs(vtkMath::Dot(right, up)) > 0.999)
{
  right = { 0.0, 1.0, 0.0 };
}
```

Reply:
true, I reused the code from a previous version of the function where `right` was a param, it's now hardcoded so we can simplify